feat: [005-Post-Search] Elasticsearch를 통한 게시글 검색 기능 구현

jaehyeok-code opened the pull request:
Changes
---
- 부분문자열 검색을 위해 n-gram-analyzer를 적용
- docker-compose.yml에 Elasticsearch 설정 추가

jaehyeok-code committed:
feat: [005-Post-Search] Elasticsearch를 통한 게시글 검색 기능 구현

- 부분문자열 검색을 위해 n-gram-analyzer를 적용
- docker-compose.yml에 Elasticsearch 설정 추가

mingming-mentor commented on post-api/src/main/java/com/project/post/service/PostService.java:
```diff
@@ -58,7 +60,9 @@ public Post createPost(String token, PostRequest request, MultipartFile file) {
 
-    return postRepository.save(post);
+    Post createdPost = postRepository.save(post);
+    postSearchService.indexPost(createdPost);
```
기획적으로 허용이 필요한 부분이긴 하지만, indexPost는 비동기로 처리해도 좋을 것 같네요. 
그러면 검색 시스템(es)에 문제가 생겼을때 post save자체가 안되는 문제는 발생하지 않을듯해요.
대신 약간의 지연이 있을 수 있는데 이 부분이 기획적으로 검토가 될 필요는 있겠구요

jaehyeok-code replied:
보통 커뮤니티 플랫폼이나 sns에서는 비동기 처리로 진행하는지 몰랐습니다!  제 생각엔 게시글 검색이 저장 시점 이후로 바로 되어야 하는만큼 급한 기능은 아니라고 생각해서 일단 비동기 처리로 진행하겠습니다!

mingming-mentor commented on post-api/src/main/java/com/project/post/search/PostSearchController.java:
```diff
@@ -0,0 +1,46 @@
+
+    jwtAuthenticationProvider.getUserVo(token);
+
+    Page<PostDocument> esPage = postSearchService.searchPosts(query, page, size);
+    List<PostResponse> responses = esPage.getContent().stream()
+        .map(doc -> PostResponse.builder()
+            .id(doc.getId())
+            .userId(doc.getUserId())
+            .title(doc.getTitle())
+            .content(doc.getContent())
+            .createdAt(doc.getCreatedAt())
+            .updatedAt(doc.getUpdatedAt())
+            .build())
+        .toList();
```
esPage는 별도로 선언하지 않아도 될것 같습니다. 
```suggestion
    List<PostResponse> responses = postSearchService.searchPosts(query, page, size).getContent().stream()
        .map(doc -> PostResponse.builder()
            .id(doc.getId())
            .userId(doc.getUserId())
            .title(doc.getTitle())
            .content(doc.getContent())
            .createdAt(doc.getCreatedAt())
            .updatedAt(doc.getUpdatedAt())
            .build())
        .toList();
```

jaehyeok-code replied:
네 확인해봤는데 수정하겠습니다.

mingming-mentor reviewed (approved):
수고 많으셨습니다

jaehyeok-code committed:
feat: [005-Post-Search] Elasticsearch를 통한 게시글 검색 기능 구현 + 추가 수정

- 부분문자열 검색을 위해 n-gram-analyzer를 적용
- docker-compose.yml에 Elasticsearch 설정 추가
- indexpost 비동기 처리
- PostSearchController 수정
